### PR TITLE
deprecate static system names in the FEMechanicsBase, IBFEMethod, and FEDataManager interfaces

### PR DESCRIPTION
**Pull request by boyceg**
### IBAMR Pull Request Checklist
- [x] Does the test suite pass?
- [x] Was clang-format run?
- [x] Were relevant issues cited? Please remember to add `Fixes #12345` to close
      the issue automatically if we are fixing the problem.
- [x] Is this a change others will want to know about? If so, then has a
      changelog entry been added?
- [x] If new data structures have been added to a class then have they been set
      up appropriately for restarts? If so, ensure that the restart version
      number is incremented.
- [x] Does this change include a bug fix or new functionality? If so, a new test
      or tests should be added. New tests should run quickly (less than a minute
      in release mode). If possible, an older test should gain a new option so
      that we do not need to compile more test executables.
- [x] Did you (if your account has permission to do so) set relevant labels on
      GitHub for the pull request?

### CODE REVIEW

**Comment by boyceg** [comment]
Replaces #1514. (GitHub would not let me reopen that PR.)

File: ibtk/include/ibtk/EdgeDataSynchronization.h [hunk 1]
```diff
@@ -69,7 +69,8 @@ class EdgeDataSynchronization : public SAMRAI::tbox::DescribedClass
         /*!
          * \brief Default constructor.
          */
-        inline SynchronizationTransactionComponent(int data_idx = -1, const std::string& coarsen_op_name = "NONE")
+        inline SynchronizationTransactionComponent(int data_idx = invalid_index,
```
Comment (drwells) [inline ibtk/include/ibtk/EdgeDataSynchronization.h:72]: I think some other stuff got tangled up in here.

File: ibtk/include/ibtk/FEDataManager.h [hunk 2]
```diff
@@ -548,14 +548,43 @@ class FEDataManager : public SAMRAI::tbox::Serializable
...
+
 public:
     /*!
      * \brief The name of the equation system which stores the spatial position
      * data. The actual string is stored by FEData.
      *
      * \note The default value for this string is "coordinates system".
      */
-    std::string& COORDINATES_SYSTEM_NAME;
+    [[deprecated(
```
Comment (drwells) [inline ibtk/include/ibtk/FEDataManager.h:562]: These annotations are, I think, C++17: we'll need to either update to C++17 or define a compatibility macro `IBAMR_DEPRECATED("message")`

**Comment by boyceg** [comment]
This led me to believe that it was valid C++11:

https://en.cppreference.com/w/cpp/language/attributes

**Comment by boyceg** [comment]
Ah — missed that deprecated is C++14.

File: ibtk/include/ibtk/EdgeDataSynchronization.h (hunk 1 above)
Comment (boyceg) [reply ibtk/include/ibtk/EdgeDataSynchronization.h]: Looks like I messed up rebasing against `master`. That is what I get for force pushing code while adjudicating a wrestling match between my kids. I think this is fixed now.

File: ibtk/include/ibtk/FEDataManager.h (hunk 2 above)
Comment (boyceg) [reply ibtk/include/ibtk/FEDataManager.h:562]: Let's do the latter...

**Comment by boyceg** [comment]
OK, all checks are passing now --- although I had to add a bit of a hack to avoid warnings for `Build IBAMR and run tests (pull_request)`.

File: ibtk/include/ibtk/ibtk_utilities.h [hunk 3]
```diff
@@ -52,6 +52,18 @@ IBTK_ENABLE_EXTRA_WARNINGS
         }                                                                                                              \
     } while (0);
 
+#ifdef IBTK_NDEF_IBTK_DEPRECATED
+#define IBTK_DEPRECATED(msg)
+#elif defined(__cplusplus) && (__cplusplus >= 201402L)
```
Comment (boyceg) [inline ibtk/include/ibtk/ibtk_utilities.h:57]: It would probably be better if we detected whether the compiler supports `deprecated`...

Comment (drwells) [reply ibtk/include/ibtk/ibtk_utilities.h:57]: I can write a followup patch to add this to our feature detection.

**Comment by boyceg** [comment]
@drwells, does this look OK?

**Comment by drwells** [comment]
I think this is generally OK but I'd like to apply some minor patches before we merge. We can suppress warnings in a cleaner way by doing

```cpp

#define COORDS_SYSTEM_NAME_VAL "IB coordinates system"
#define COORD_MAPPING_SYSTEM_NAME_VAL "IB coordinate mapping system"
#define FORCE_SYSTEM_NAME_VAL "IB force system"
#define PRESSURE_SYSTEM_NAME_VAL "IB pressure system"
#define VELOCITY_SYSTEM_NAME_VAL "IB velocity system"

// Suppress warnings so that we can cleanly define these deprecated variables
IBTK_DISABLE_EXTRA_WARNINGS
const std::string FEMechanicsBase::COORDS_SYSTEM_NAME = COORDS_SYSTEM_NAME_VAL;
const std::string FEMechanicsBase::COORD_MAPPING_SYSTEM_NAME = COORD_MAPPING_SYSTEM_NAME_VAL;
const std::string FEMechanicsBase::FORCE_SYSTEM_NAME = FORCE_SYSTEM_NAME_VAL;
const std::string FEMechanicsBase::PRESSURE_SYSTEM_NAME = PRESSURE_SYSTEM_NAME_VAL;
const std::string FEMechanicsBase::VELOCITY_SYSTEM_NAME = VELOCITY_SYSTEM_NAME_VAL;
IBTK_ENABLE_EXTRA_WARNINGS
```
instead of messing around with the header. Would you mind if I pushed the changes onto this branch directly?

**Comment by boyceg** [comment]
> Would you mind if I pushed the changes onto this branch directly?

Fine with me!

**Review by drwells — APPROVED** [review]

File: include/ibamr/FEMechanicsBase.h [hunk 4]
```diff
@@ -59,16 +59,16 @@ namespace IBAMR
...
     static const std::string COORDS_SYSTEM_NAME;
-    IBTK_DEPRECATED("use FEMechanicsBase::getDisplacementSystemName() to access the current coordinates system name.")
+    IBTK_DEPRECATED(
+        "Use FEMechanicsBase::getVelocitySystemName() to access the current coordinate mapping system name.")
     static const std::string COORD_MAPPING_SYSTEM_NAME;
-    IBTK_DEPRECATED("use FEMechanicsBase::getForceSystemName() to access the current coordinates system name.")
+    IBTK_DEPRECATED("Use FEMechanicsBase::getVelocitySystemName() to access the current force system name.")
     static const std::string FORCE_SYSTEM_NAME;
-    IBTK_DEPRECATED("use FEMechanicsBase::getPressureSystemName() to access the current coordinates system name.")
+    IBTK_DEPRECATED("Use FEMechanicsBase::getVelocitySystemName() to access the current pressure system name.")
```
Comment (abarret) [inline include/ibamr/FEMechanicsBase.h:69]: ?
```suggestion
    IBTK_DEPRECATED("Use FEMechanicsBase::getPressureSystemName() to access the current pressure system name.")
```

File: include/ibamr/FEMechanicsBase.h [hunk 5]
```diff
@@ -59,16 +59,16 @@ namespace IBAMR
...
-    IBTK_DEPRECATED(
-        "use FEMechanicsBase::getCurrentCoordinatesSystemName() to access the current coordinates system name.")
+    IBTK_DEPRECATED("Use FEMechanicsBase::getVelocitySystemName() to access the current coordinates system name.")
     static const std::string COORDS_SYSTEM_NAME;
-    IBTK_DEPRECATED("use FEMechanicsBase::getDisplacementSystemName() to access the current coordinates system name.")
+    IBTK_DEPRECATED(
+        "Use FEMechanicsBase::getVelocitySystemName() to access the current coordinate mapping system name.")
     static const std::string COORD_MAPPING_SYSTEM_NAME;
-    IBTK_DEPRECATED("use FEMechanicsBase::getForceSystemName() to access the current coordinates system name.")
+    IBTK_DEPRECATED("Use FEMechanicsBase::getVelocitySystemName() to access the current force system name.")
```
Comment (abarret) [inline include/ibamr/FEMechanicsBase.h:67]: ?
```suggestion
    IBTK_DEPRECATED("Use FEMechanicsBase::getForceSystemName() to access the current force system name.")
```

File: include/ibamr/FEMechanicsBase.h (hunk 4 above)
Comment (boyceg) [reply include/ibamr/FEMechanicsBase.h:69]: Did these get garbled somewhere along the way?

File: include/ibamr/FEMechanicsBase.h (hunk 5 above)
Comment (drwells) [reply include/ibamr/FEMechanicsBase.h:67]: Whoops

File: ibtk/include/ibtk/FEDataManager.h [hunk 6]
```diff
@@ -555,8 +555,33 @@ class FEDataManager : public SAMRAI::tbox::Serializable
      *
      * \note The default value for this string is "coordinates system".
      */
+    IBTK_DEPRECATED(
+        "use FEDataManager::getCurrentCoordinatesSystemName() and FEDataManager::setCurrentCoordinatesSystemName() to "
```
Comment (boyceg) [inline ibtk/include/ibtk/FEDataManager.h:559]: @drwells your last commit changed some of these messages from `"use..."` to `"Use..."`. Do you want to make this consistent across the library?

Comment (drwells) [reply ibtk/include/ibtk/FEDataManager.h:559]: Good thing we have some QA :) I'll fix this.

Comment (boyceg) [reply ibtk/include/ibtk/FEDataManager.h:559]: I don't care that much --- I just noticed that you changed this (from what had previously been the consistent use of lowercase 😁)

File: include/ibamr/FEMechanicsBase.h [hunk 7]
```diff
@@ -113,6 +116,46 @@ class FEMechanicsBase : public SAMRAI::tbox::Serializable
      */
     libMesh::EquationSystems* getEquationSystems(unsigned int part = 0) const;
 
+    /*!
+     * Return the current coordinates system name.
+     */
+    inline const std::string& getCurrentCoordinatesSystemName() const
```
[the remark this answers is not in the document]
Comment (boyceg) [reply include/ibamr/FEMechanicsBase.h:122]: Do you want to drop them, or shall I?

[the remark this answers is not in the document]
Comment (drwells) [reply include/ibamr/FEMechanicsBase.h:122]: I have a patch to push but I want to check one more thing

[the remark this answers is not in the document]
Comment (boyceg) [reply include/ibamr/FEMechanicsBase.h:122]: Oof.

**Comment by drwells** [comment]
If I have time I'll investigate the HYPRE bug - things definitely work with 2.23 and segfault with 2.26. Regardless - I rebased and this should be good to go.

**Comment by boyceg** [comment]
> If I have time I'll investigate the HYPRE bug - things definitely work with 2.23 and segfault with 2.26. Regardless - I rebased and this should be good to go.

I'd like to take one more look at the changes before merging.

**Comment by drwells** [comment]
Any final thoughts or can we put this in 0.12?

**Comment by boyceg** [comment]
I would like to put this in. Can you take a quick look to make sure the old/new names appear to match? That is the main thing I was going to do, but I’m not sure when I will be able to get to it.

**Comment by drwells** [comment]
Sure - I'll double-check and merge.

**Review by drwells — APPROVED** [review]
This PR looks good (and it includes two other minor fixes). I checked and the heart model still runs - I don't see any reason why other external codes would stop working.

File: src/IB/IBFEMethod.cpp [hunk 8]
```diff
@@ -603,8 +615,6 @@ IBFEMethod::preprocessIntegrateData(double current_time, double new_time, int nu
     }
     if (d_Q_vecs) d_Q_vecs->copy("solution", { "current", "half" });
 
-    // Update the mask data.
-    getVelocityHierarchyDataOps()->copyData(mask_new_idx, mask_current_idx);
```
[the remark this answers is not in the document]
Comment (drwells) [reply src/IB/IBFEMethod.cpp:607]: Like the other one - we don't ever use these so lets get rid of them.

**Pull request merged**